contrib/guix/libexec/build.sh

```diff
@@ -171,6 +171,8 @@ make -C depends --jobs="$JOBS" HOST="$HOST" \
                                    ${SOURCES_PATH+SOURCES_PATH="$SOURCES_PATH"} \
                                    ${BASE_CACHE+BASE_CACHE="$BASE_CACHE"} \
                                    ${SDK_PATH+SDK_PATH="$SDK_PATH"} \
+                                   native_qt_cc="${NATIVE_GCC}/bin/gcc" \
```

Comment:
In 57b8139a9c146f1146b9d3c92be35a7fb606585f (guix: Adjust for Qt 6): Not sure this is the right direction.
1. Why does Qt need special treatment here? It should be finding the compiler, the same as all other packages we build.
2. We shouldn't be hard-coding GCC usage, otherwise you're requiring a GCC toolchain to build releases for macOS, where it shouldn't be needed. Clang is capable of building native bins. i.e #30206.

Reply:
1. On the master branch, building in Guix for macOS does use a GCC toolchain to compile native Qt tools.

This PR does the same.

I agree that Clang could be used instead. However, I'd prefer to leave this change of of this PR scope.

2. Special treatment is needed for every native package that uses CMake. For now, this applies only to Qt.

By default, CMake looks for the `cc` and `c++` executables available in `PATH`.  A GCC toolchain creates `bin/gcc` and `bin/c++` symlinks, while a Clang toolchain creates `bin/cc` and `bin/c++`, respectively. Therefore, CMake can end up with the following configuration:
```
$ cat CMakeLists.txt 
cmake_minimum_required(VERSION 3.22)
project(guix-play)

$ guix shell --container --pure cmake ninja clang-toolchain@18 gcc-toolchain@12
[env]$ cmake -B build -G Ninja --fresh
-- The C compiler identification is Clang 18.1.8
-- The CXX compiler identification is GNU 12.4.0
< snip >
```
This configuration fails to build on `x86_64`.

Reply:
> Special treatment is needed for every native package that uses CMake. For now, this applies only to Qt.

So this will also apply to libmultiprocess. I guess it's currently not a issue?, because we don't build it in Guix, however, that may change soon, #30975, so rather than adding Qt workarounds here, and then having to either add the same workarounds for libmultiprocess, or fix things in a more fundamental way later, it may still be better to make those changes now. 

> By default, CMake looks for the cc and c++ executables available in PATH. A GCC toolchain creates bin/gcc and bin/c++ symlinks, while a Clang toolchain creates bin/cc and bin/c++, respectively. Therefore, CMake can end up with the following configuration:

GCC and Clang installs may create those symlinks in some default cases, but this will all vary by install or distro, for example, on Alpine:
```bash
apk add gcc clang
# cc --version
cc (Alpine 13.2.1_git20240309) 13.2.1 20240309 # cc here is GCC, not Clang
```

So it'd seem like this is an issue that can already happen outside of Guix, which might mean it's better to fix it in depends, rather than adding workarounds for depends (just inside Guix).

Note that in the Guix example you provided, whether you see that behaviour, depends on the install-order of the packages. i.e if `clang-toolchain` is installed last, it will be `cc` & `c++`, if `gcc-toolchain` is last, you'll have `cc` as Clang, and `c++` will be GCC.

Reply:
Please disregard my previous comments.

I've reviewed and tested the current branch once more. Modifications in `contrib/guix/libexec/build.sh` are not required. Setting proper compilers is handled already:
- for the `native_qt` package: https://github.com/bitcoin/bitcoin/blob/af05dd9a12b89224dc7ad229698eeceb3e560ed4/depends/packages/native_qt.mk#L88-L89
- for other CMake-based native packages: https://github.com/bitcoin/bitcoin/blob/af05dd9a12b89224dc7ad229698eeceb3e560ed4/depends/funcs.mk#L178-L181

Will update this PR shortly.